### DISCUSSION TITLE
How to check foreign keys?

### DISCUSSION THREAD

**Discussion by JenuelDev** [laravel/framework#43443] [Q&A] [answered]
From laravel 4 to 9. It seems there has been no checker for the schema to check foreign keys. I was hoping from v4 to 9 to have something like this. But it seems until now, there is no function. I need help,

The Function I has hoping for:
```php
Schema::hasForeign('<table name>', ['<column name>']); // that will return true or false
```

**Comment by henzeb** [comment]
It's a bit hacky, but this may be a solution.

```php
$doctrineTable = Schema::getConnection()->getDoctrineSchemaManager()->listTableDetails('table_name');

$doctrineTable->hasForeignKey('foreignKeyName');
```

**Reply by JenuelDev** [reply]
thanks mate!, that is what I did, but I have to go over my database and check for the key.. it would be awesome if I can just add the column name and automatically check the foreign key.,

**Comment by henzeb** [comment, marked as answer]
With `getForeignKeyColumns()` you can get an array of columns who are involved in a foreign key constraint.

With `getForeignKeys()` you get an arrray of `ForeignKeyConstraint` which has a method `getLocalColumns` with an array you can use to  check if the column you want is involved.  

I agree it's a little too much for something so simple, and we cannot add macro's to the Schema builder.

**Reply by megabyte-0** [reply]
In laravel 11, where the `getDoctrineSchemaManager` is no longer available in `\Illuminate\Database\Schema\Builder`, there is [`Schema::getForeignKeys($table)`](https://github.com/laravel/framework/blob/4b5ce19b3bbe1d317585d4b98da66fbb6e7357af/src/Illuminate/Support/Facades/Schema.php#L28).  
I use it like this:
```php
    protected array $foreignKeyCache = [];

    protected function getForeignKeyByColumns(string $table, array $columns): ?array
    {
        $delimiter = '|';
        $glue = fn($columns) => collect($columns)->join($delimiter);
        if (! isset($this->foreignKeyCache[$table])) {
            $this->foreignKeyCache[$table] = collect(Schema::getForeignKeys($table))->keyBy(
                fn (array $foreignKeyItemRecord) => $glue($foreignKeyItemRecord['columns'])
            )->toArray();
        }
        return $this->foreignKeyCache[$table][$glue($columns)] ?? null;
    }
```
